src/tools/wasm-link.c

```diff
@@ -0,0 +1,909 @@
+      wasm_write_u32_leb128(stream, sig_index, "sig");
+    }
+  }
+
+  FIXUP_SIZE;
+}
+
+static void write_data_segment(WasmStream* stream,
+                               DataSegment* segment,
+                               uint32_t offset) {
+  assert(segment->memory_index == 0);
+  wasm_write_u32_leb128(stream, segment->memory_index, "memory index");
+  wasm_write_opcode(stream, WASM_OPCODE_I32_CONST);
+  wasm_write_u32_leb128(stream, segment->offset + offset, "offset");
+  wasm_write_opcode(stream, WASM_OPCODE_END);
+  wasm_write_u32_leb128(stream, segment->size, "segment size");
+  wasm_write_data(stream, segment->data, segment->size, "segment data");
+}
+
+static void write_combined_data_section(Context* ctx,
+                                        SectionPtrVector* sections,
+                                        uint32_t total_count) {
+  WasmStream* stream = &ctx->stream;
+  WRITE_UNKNOWN_SIZE;
+
+  wasm_write_u32_leb128(stream, total_count, "element count");
+  size_t i, j;
+  for (i = 0; i < sections->size; i++) {
+    Section* sec = sections->data[i];
+    for (j = 0; j < sec->data_segments.size; j++) {
+      DataSegment* segment = &sec->data_segments.data[j];
+      write_data_segment(stream, segment,
+                         sec->binary->memory_page_offset * WASM_PAGE_SIZE);
+    }
+  }
+
+  FIXUP_SIZE;
+}
+
+static void write_combined_reloc_section(Context* ctx,
+                                         WasmBinarySection type,
+                                         SectionPtrVector* sections,
+                                         uint32_t section_index) {
+  size_t i, j;
+  uint32_t total_relocs = 0;
+
+  // First pass to know total reloc count
+  for (i = 0; i < sections->size; i++) {
+    Section* sec = sections->data[i];
+    total_relocs += sec->relocations.size;
+  }
+
+  if (!total_relocs)
+    return;
+
+  char section_name[128];
+  sprintf(section_name, "%s.%s", WASM_BINARY_SECTION_RELOC,
+          wasm_get_section_name(type));
+
+  WasmStream* stream = &ctx->stream;
+  wasm_write_u8(stream, WASM_BINARY_SECTION_CUSTOM, "section code");
+  WRITE_UNKNOWN_SIZE;
+  wasm_write_str(stream, section_name, strlen(section_name),
+                 WASM_PRINT_CHARS, "reloc section name");
+  wasm_write_u32_leb128(&ctx->stream, section_index, "reloc section");
+  wasm_write_u32_leb128(&ctx->stream, total_relocs, "num relocs");
+
+  for (i = 0; i < sections->size; i++) {
+    Section* sec = sections->data[i];
+    RelocVector* relocs = &sec->relocations;
+    for (j = 0; j < relocs->size; j++) {
+      wasm_write_u32_leb128(&ctx->stream, relocs->data[j].type, "reloc type");
+      uint32_t new_offset = relocs->data[j].offset + sec->output_payload_offset;
+      wasm_write_u32_leb128(&ctx->stream, new_offset, "reloc offset");
+    }
+  }
+
+  FIXUP_SIZE;
+}
+
+static WasmBool write_combined_section(Context* ctx,
+                                       WasmBinarySection type,
```

Comment:
nit: I've been calling this section_code elsewhere, since type is pretty overloaded (though section_code is a little confusing too, considering there is a code section...)